docs/design/features/source-generator-com.md

```diff
@@ -0,0 +1,659 @@
+    }
+
+    private class ComObject : IDynamicInterfaceCastable
+    {
+        private IntPtr iMyComInterface;
+        private IntPtr iMyOtherComInterface;
+        private IntPtr iDerivedComInterface;
+
+        public ComObject(IntPtr externalComObject)
+        {
+            // ...
+        }
+    }
+
+    [DynamicInterfaceCastableImplementation]
+    private interface IMyComInterfaceImpl : IMyComInterface
+    {
+        // ...
+    }
+
+    [DynamicInterfaceCastableImplementation]
+    private interface IMyOtherComInterfaceImpl : IMyOtherComInterface
+    {
+        // ...
+    }
+
+    [DynamicInterfaceCastableImplementation]
+    private interface IMyDerivedComInterfaceImpl : IMyDerivedComInterface
+    {
+        // ...
+    }
+
+    public static readonly MyComWrappers Instance = new();
+
+    public struct Marshaller<T>
+    {
+
+    }
+}
+```
+
+Pros:
+
+- Can use de-compiled TlbImp-generated COM Interop code with minimal changes.
+- Can be designed such that no new types are required to ship with the generator or that the generator must add to the compilation.
+
+Cons:
+
+- Some changes from the TlbImp-generated code would still be required to avoid falling into the built-in system.
+- We wouldn't be able to easily fix the warts of the existing system because that would break possible back-compat.
+- The runtime might have cases where it makes assumptions based on how types are marked that this may interact with.
+- It would be much easier to accidentally use the old system and get in a place where 2 objects, one from the built-in system and one from the generated system, both represent the same native object.
+
+#### Option 5: Only `GeneratedComImport` attribute with Generated `ComWrappers`-derived type
+
+The built-in `ComImport` and `ComVisible` attributes have a lot of history and weird runtime behavior associated with them. Additionally the built-in `ComVisible` attribute actually takes a `bool` to determine if the applied to type is visible and it can be applied to methods as well to enable/disable COM visbility for the legacy automatic COM vtable generation that the .NET runtime has supported since .NET Framework 1.0. This option proposes introducing a single new attribute to cover the expected scenarios:
+
+```csharp
+[AttributeUsage(AttributeTargets.Interface)]
+class GeneratedComInterfaceAttribute
+{
+    public GeneratedComInterfaceAttribute();
+
+    public GeneratedComInterfaceAttribute(bool generateManagedObjectWrapper, bool generateComObjectWrapper);
+
+    public bool GenerateManagedObjectWrapper { get; } = true;
+
+    public bool GenerateComObjectWrapper { get; } = true;
+
+    public bool ExportMetadata { get; }
+}
+```
+
+This attribute could be applied to any interface to generate the marshalling code for either RCW or CCW scenarios. The `ExportMetadata` property would be used by any tool that wants to generate a metadata file, like a TlbExp successor, to determine which interfaces to export.
```

Comment:
> The `ExportMetadata` property

It might be better to be explicit about what the "metadata" format. Based on some offline conversations it would seem IDL is what is going to be the source generator's artifact. From that, a TLB could be generated if needed.

Also, I just realized that expecting this to be a TLB is going to make the cross-platform scenario weird since the attribute would mean nothing without us implementing `CreateTypeLib()` for non-Windows... blech.